### PR TITLE
Modularize Bazel build

### PR DESCRIPTION
**Pull request by mofeing**
Working on #370, i realized it was maybe time to refactor the build and make it more modular, like other Bazel-based projects have.

This PR moves the dependencies to different files in a modular-manner. This organization also works if we move to higher versions of Bazel, where WORKSPACE is deprecated and Bazelmod must be used.

- `BUILD` should most probably be empty, but still required so that Bazel recognizes that folder as a package (like `__init__.py` in Python).
- `workspace.bzl` defines a `repo` function that imports the repository. This `repo` function is then called from `WORKSPACE`.
- `*.BUILD` in the case of local repositories or repos which are not Bazel-build, this file should wrap it to make Bazel-compatible. check out #370 how i did it for Julia and libcxxwrap_julia.

Btw, I've also started modularizing the sources of ReactantExtra into different files. The reason is that currently we only have 1 big file, "API.cpp", that if we divide into different files, compile time gets reduced a lot for the files involving ReactantExtra. It's nice for development. For example, when moving the IFRT part to another file (still wip), compile time (caching all XLA and MLIR deps) reduced from 60s to 30s in my laptop.

### CODE REVIEW

File: deps/ReactantExtra/third_party/hedron_compile_commands/workspace.bzl [hunk 1]
```diff
@@ -0,0 +1,15 @@
+"""Loads Hedron's Compile Commands Extractor for Bazel."""
+# https://github.com/hedronvision/bazel-compile-commands-extractor
+
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
```
[the remark this answers is not in the document]
Comment (mofeing) [reply deps/ReactantExtra/third_party/hedron_compile_commands/workspace.bzl:4]: do you mean importing this rule from enzyme-jax itself? not 100% how can it be done right now as Enzyme-JAX is (i.e. without the modular deps), but i guess it doesn't matter much if we leave it like it is and remove it in the future once we "modularize" the Enzyme-JAX build

[the remark this answers is not in the document]
Comment (wsmoses) [reply deps/ReactantExtra/third_party/hedron_compile_commands/workspace.bzl:4]: yeah fair nuff, we should probably fix that up then use the corresponding imports here

[the remark this answers is not in the document]
Comment (mofeing) [reply deps/ReactantExtra/third_party/hedron_compile_commands/workspace.bzl:4]: okay, lemme do the same kind of PR in Enzyme-JAX

**Comment by mofeing** [comment]
The PR for the modularization of the Bazel deps in Enzyme-JAX is open now. Once it's merged, I'll refactor this PR on top of it.

**Comment by mofeing** [comment]
@wsmoses this PR should be ready. do you want to move the commit vars to a `workspace.bzl`, just like in Enzyme-JAX?

**Comment by wsmoses** [comment]
Yeah just because we will be constantly doing version bumps doing this all in one place will be best

**Comment by mofeing** [comment]
done

**Comment by wsmoses** [comment]
sorry @mofeing I'm going to ask that we merge https://github.com/EnzymeAD/Reactant.jl/pull/427 first since that contains jll changes that have been already built in the last jll in yggdrassil but haven't been merged (and we should merge as is and then rebase on top of it, otherwise we won't have a way to rebuild this present jll if need be)

**Review by wsmoses — APPROVED** [review]
LGTM but yeah we need to land https://github.com/EnzymeAD/Reactant.jl/pull/427 first [which hopefully happens tonight] per reason above

**Comment by mofeing** [comment]
okay, no problem. i will wait for it to land.

just ping me whenever is ready.

**Pull request merged**